### DISCUSSION TITLE
NFT crafting/combining: Where to store `burnRef`?

### DISCUSSION THREAD

**Discussion by arjanjohan** [aptos-labs/aptos-developer-discussions#381] [Questions] [answered]
### Discord user ID

arjanjohan

### Describe your question in detail.

tldr: How to burn NFT's? Any code examples of NFT's that are used (to combine/craft into new NFT's) in Aptos Move?

We are working on an NFT platform where users can combine 2 NFT's and create a new NFT. The old NFT's should get burned. We started with the `nft_launchpad.move` file as template. But we are having issues with burning the old NFT's. Where should we store the `BurnRef`?

Currently I added this. When minting an NFT, the `BurnRef` is stored here.
```
    /// Unique per collection
    struct CollectionBurnRefs has key, store, drop {
        burn_ref_map: SimpleMap<address, token::BurnRef>,
    }
```

But when I actually want to burn the NFT's, I cannot dereference the `BurnRef`and use it. See my 

```
        // Retrieve the burn_refs

        let main_nft_addr = object::object_address(&main_nft);
        let other_nft_addr = object::object_address(&other_nft);

        let collection_burn_refs = borrow_global<CollectionBurnRefs>(object::object_address(&main_collection_obj));
        let main_nft_burn_ref = simple_map::borrow(&collection_burn_refs.burn_ref_map, &main_nft_addr);
        let other_nft_burn_ref = simple_map::borrow(&collection_burn_refs.burn_ref_map, &other_nft_addr);

        // Burn NFTs using the retrieved burn_refs
        token::burn(*main_nft_burn_ref);
        token::burn(*other_nft_burn_ref);
```

Source code:
https://github.com/gmove-hackers/gmove-nft/blob/move-modules/move/sources/launchpad.move

### What error, if any, are you getting?

error[E05001]: ability constraint not satisfied
    ┌─ /home/arjan/git/hack/olympus/gmove/move/sources/launchpad.move:401:21
    │
116 │         burn_ref_map: SimpleMap<address, token::BurnRef>,
    │                                          -------------- The type '(aptos_token_objects=0x4)::token::BurnRef' does not have the ability 'copy'
    ·
401 │         token::burn(*other_nft_burn_ref);
    │                     ^^^^^^^^^^^^^^^^^^^ Invalid dereference. Dereference requires the 'copy' ability
    │
    ┌─ /home/arjan/.move/https___github_com_aptos-labs_aptos-core_git_mainnet/aptos-move/framework/aptos-token-objects/sources/token.move:93:12
    │
 93 │     struct BurnRef has drop, store {
    │            ------- To satisfy the constraint, the 'copy' ability would need to be added here

### Which operating system are you using?

Linux (Ubuntu, Fedora, Windows WSL, etc.)

### Which SDK or tool are you using? (if any)

N/A

**Comment by gregnazario** [comment, marked as answer]
The easiest way is to put the `BurnRef` in the token's object itself with a `move_to` with the object's address.  You will then need to do a `move_from` the object's address to recover the `BurnRef`, which must be consumed by `burn`.  You cannot use a reference of the `BurnRef`

Here's an example https://github.com/aptos-labs/daily-move/blob/main/snippets/modifying-nfts/sources/modify_nfts.move#L201-L228

**Comment by arjanjohan** [comment]
Great, that is exactly what I needed. I adjusted the code according to the example you gave and it compiles. Thanks for unblocking us! 
Also  the other daily-move examples seem like they will be very useful for us!
